csp/impl/perspective_common.py

```diff
@@ -0,0 +1,130 @@
+import pytz
+from datetime import date, datetime, timedelta
+from packaging import version
+
+import csp
+
+try:
+    import perspective
+
+    if version.parse(perspective.__version__) >= version.parse("3"):
+        _PERSPECTIVE_3 = True
+        from perspective.widget import PerspectiveWidget
+
+    elif version.parse(perspective.__version__) >= version.parse("0.6.2"):
+        from perspective import PerspectiveManager, PerspectiveWidget  # noqa F401
+
+        _PERSPECTIVE_3 = False
+    else:
+        raise ImportError("perspective adapter requires 0.6.2 or greater of the perspective-python package")
+
+except ImportError:
+    raise ImportError(
+        "perspective must be installed to use this module. " "To install, run 'pip install perspective-python'."
+    )
+
+
+def is_perspective3():
+    """Whether the perspective version is >= 3"""
+    return _PERSPECTIVE_3
+
+
+def perspective_type_map():
+    """Return the mapping of standard python types to perspective types"""
+    return {
+        str: "string",
+        float: "float",
+        int: "integer",
+        date: "date",
+        datetime: "datetime",
+        bool: "boolean",
+    }
+
+
+def datetime_to_perspective(dt: datetime) -> int:
+    """Convert a python datetime to an integer number of milliseconds for perspective"""
+    return int(dt.timestamp() * 1000)
+
+
+def date_to_perspective(d: date) -> int:
+    """Convert a python date to an integer number of milliseconds for perspective"""
+    return int(datetime(year=d.year, month=d.month, day=d.day, tzinfo=pytz.UTC).timestamp() * 1000)
+
+
+@csp.node
+def apply_updates(
+    table: perspective.Table,
+    data: {object: csp.ts[object]},
+    index_col: str,
+    time_col: str,
+    throttle: timedelta,
+    localize: bool,
+    static_records: dict = None,
+):
+    with csp.alarms():
+        alarm = csp.alarm(bool)
+
+    with csp.state():
+        s_buffer = []
+        s_has_time_col = False
+        s_datetime_cols = set()
+        s_date_cols = set()
+
+    with csp.start():
+        if throttle > timedelta(0):
+            csp.schedule_alarm(alarm, throttle, True)
+        s_has_time_col = time_col and time_col not in data.keys()
+        if _PERSPECTIVE_3:
+            s_datetime_cols = set([c for c, t in table.schema().items() if t == "datetime"])
+            s_date_cols = set([c for c, t in table.schema().items() if t == "date"])
+        else:
+            s_datetime_cols = set([c for c, t in table.schema().items() if t == datetime])
+
+    with csp.stop():
+        if len(s_buffer) > 0:
+            table.update(s_buffer)
+
+    if csp.ticked(data):
+        new_rows = {}
+        for (idx, col), value in data.tickeditems():
+            if idx not in new_rows:
+                row = {}
+                new_rows[idx] = row
+                if index_col:
+                    row[index_col] = idx
+                if s_has_time_col:
+                    if localize or _PERSPECTIVE_3:
```

Comment:
do we need to check both here or can we just pass `(localize or _PERSPECTIVE_3)` when calling the node

Reply:
I'll leave itt o you, but just saying it looks like all checks being done are the same `localize or _PERSPECTIVE_3` so may as well lift it out of the node

Reply:
I've refactored this so that the right handling for perspective 3 is just done inside the datetime conversion function